Add InvokeAIAppConfig schema migration system

lstein opened the pull request:
## Summary

This implements a straightforward system for migrating the `invokeai.yaml` configuration file between `InvokeAIAppConfig` schema versions.

To illustrate how it works, say that we are currently at config schema version 4.0.1, which includes an `attention_type` value `xformers`. But `xformers` becomes unsupported and we need to remove it. So we remove this choice from the list of valid values for `attention_type`, and update `CONFIG_SCHEMA_VERSION` to `4.0.2`. But some users will have an old `invokeai.yaml` that contains `attention_type: xformers` and if they try to run the app it will fail at startup time with a pydantic schema validation error.

Here's what we do to fix that. In `invokeai/app/services/config/migrations.py` we add the following schema migration function to the `load()` method:

```
@migrator.register(from_version="4.0.1", to_version="4.0.2")
def migrate_3(config_dict: AppConfigDict) -> AppConfigDict:
   new_config_dict: AppConfigDict = {}
   for k, v in config_dict.items():
        if k == "xformers":
              continue
        else:
              new_config_dict[k] = v
   return new_config_dict
```

The `migrator.register()` decorator is defined in `invokeai/app/services/config/config_migrate.py`. You provide it with parameters indicating the incoming and outgoing version numbers, and define a function that takes a dictionary dump from the incoming schema and returns a dictionary compatible with the outgoing schema.

The function that calls this takes care of inserting the `schema_version` field and validating against the current schema after all the migrations are completed.

When `4.0.3` comes around, you add a similar function to `config_default.py` using the decorator `@migrator.register(from_version="4.0.2", to_version="4.0.3")`

<!--A description of the changes in this PR. Include the kind of change (fix, feature, docs, etc), the "why" and the "how". Screenshots or videos are useful for frontend changes.-->

## Related Issues / Discussions

<!--WHEN APPLICABLE: List any related issues or discussions on github or discord. If this PR closes an issue, please use the "Closes #1234" format, so that the issue will be automatically closed when the PR merges.-->

## QA Instructions

First test whether it messes up your current up to date `invokeai.yaml` file (it shouldn't) by launching the web app. The code makes a backup in `invokeai.yaml.bak`, so you'll have  backup copy just in case.

Next create an older version of invokeai.yaml. I have tested with an old 3.7-era file. Run the app. It should load fine, and `invokeai.yaml` will be brought up to date, preserving any customized settings from the 3.7 file.

<!--WHEN APPLICABLE: Describe how we can test the changes in this PR.-->

## Merge Plan

Squash merge when approved.

<!--WHEN APPLICABLE: Large PRs, or PRs that touch sensitive things like DB schemas, may need some care when merging. For example, a careful rebase by the change author, timing to not interfere with a pending release, or a message to contributors on discord after merging.-->

## Checklist

- [X] _The PR has a short but descriptive title, suitable for a changelog_
- [X] _Tests added / updated (if applicable)_
- [X] _Documentation added / updated (if applicable)_

psychedelicious commented on invokeai/app/services/config/config_default.py:
```diff
@@ -512,3 +437,73 @@ def get_config() -> InvokeAIAppConfig:
+####################################################
+
+
+@ConfigMigrator.register(from_version="0.0.0", to_version="4.0.0")
```
In the migrator logic, we set the version to "3.0.0" if the schema version is missing and we are on a v3 config file, but here we register the migration as if it was from "0.0.0". This seems off

lstein replied:
Changed to 3.0.0 and tested.

psychedelicious reviewed (changes requested):
While reviewing I cleaned up and simplified a few things.

I noticed that the migrator looks for a version between a range to decide which migration to run.

I think we should be very strict here, more like the sqlite migrator class. Migrations work migrate data from an explicit version to one other explicit version, and any other versions result in an error.

I added a failing test case that should pass with stricter migration version logic. This test fails on `main` too, but I'd like this PR to improve on the current situation.

lstein commented:
> I added a failing test case that should pass with stricter migration version logic. This test fails on `main` too, but I'd like this PR to improve on the current situation.

@psychedelicious Thanks for the tidying. I've made the migration logic strict by checking for contiguous ranges in the decorator parameters before any migrations begin.

In the new test case, I don't understand why this config is expected to fail:

```
invalid_v4_0_1_config = """
schema_version: 4.0.1

host: "192.168.1.1"
port: 8080
"""
```
It looks OK to me. Perhaps it was intended to fail when the config file version was 4.0.0? I've changed the port number to "ice cream", and the test correctly raises an invalid schema error now.

lstein commented:
@psychedelicious  I updated the `test_deny_nodes()` test, which was marked as `xfailed`. It seems to work as expected now, both in isolation and as part of the full test suite. It looks like there used to be schema validation tests in the `Graph` class, but I don't see them in there any longer, so I removed the Graph assertions and kept the `BaseInvocation.get_invocations()` assertions.

psychedelicious commented:
> In the new test case, I don't understand why this config is expected to fail:

This should fail bc 4.0.1 isn't a valid registered config schema version

psychedelicious commented on tests/test_config.py:
```diff
@@ -265,50 +285,10 @@ def test_get_config_writing(patch_rootdir: None, monkeypatch: pytest.MonkeyPatch
-
-    from invokeai.app.invocations.baseinvocation import BaseInvocation
+def test_deny_nodes():
+    config = get_config()
+    config.allow_nodes = ["integer", "string", "float"]
+    config.deny_nodes = ["float"]
```
This will change the global config, lets instead instantiate a new config and let the test use that

lstein replied:
I was looking for a way to instantiate a new config and pass it to `get_invocations()` , but since this method gets its config from `get_config()` I couldn't figure out a non-hacky way of passing the custom config without altering the global.  My solution was to place a call to `get_config.cache_clear()` at the end of the test to avoid this, but I guess if the test fails halfway it will contaminate the global config. Perhaps a try/finally or context manager would work?

psychedelicious replied:
Yeah, that would work - but also I think we could just instantiate a fresh config:

```py
config = DefaultInvokeAIAppConfig()
config.allow_nodes = ["integer", "string", "float"]
config.deny_nodes = ["float"]
```

psychedelicious replied:
Sorry, you are correct. I think we need to use clear_cache.

psychedelicious commented:
> It looks like there used to be schema validation tests in the Graph class, but I don't see them in there any longer, so I removed the Graph assertions and kept the BaseInvocation.get_invocations() assertions.

Those assertion should still be in the test. They test for instantiating a graph with a denied node.

lstein commented:
> > It looks like there used to be schema validation tests in the Graph class, but I don't see them in there any longer, so I removed the Graph assertions and kept the BaseInvocation.get_invocations() assertions.
> 
> Those assertion should still be in the test. They test for instantiating a graph with a denied node.

I looked at the `Graph` class and I don't see where the test for denied nodes is occurring. The only place where I can find `app_config.deny_nodes` is in `BaseInvocation.get_invocations()`, which as far as I can tell is not called in the `Graph` constructor. Also, the Graph test fails even when I know that `config.deny_nodes` is returning the correct value.

lstein commented:
> > In the new test case, I don't understand why this config is expected to fail:
> 
> This should fail bc 4.0.1 isn't a valid registered config schema version

I'm still not following; maybe it's too late at night. The current schema version in main (and this PR) is 4.0.1 and there are two migrators defined, one from 3.0.0 to 4.0.0 and a second from 4.0.0 to 4.0.1.

psychedelicious commented:
> I'm still not following; maybe it's too late at night. The current schema version in main (and this PR) is 4.0.1 and there are two migrators defined, one from 3.0.0 to 4.0.0 and a second from 4.0.0 to 4.0.1.

Oh sorry, that's my mistake. I thought the migration you added was 4.1.0.

My intention was to test having config schema version that is between two schema versions. In the first iteration of this PR, the migrator would have treated 4.0.1 the same as 4.0.0 and migrated it to 4.1.0. Hope that makes sense.

psychedelicious commented:
> I looked at the Graph class and I don't see where the test for denied nodes is occurring. The only place where I can find app_config.deny_nodes is in BaseInvocation.get_invocations(), which as far as I can tell is not called in the Graph constructor. Also, the Graph test fails even when I know that config.deny_nodes is returning the correct value.
> 
> I'll put the Graph assertions back in. Let me know how to get them to work.

It's automatic via pydantic's union eval. I attempted to fix the root issue in #5893 but ran into what appears to be a pytest scoping issue and haven't tried to fix it again yet

lstein commented:
> > I'm still not following; maybe it's too late at night. The current schema version in main (and this PR) is 4.0.1 and there are two migrators defined, one from 3.0.0 to 4.0.0 and a second from 4.0.0 to 4.0.1.
> 
> Oh sorry, that's my mistake. I thought the migration you added was 4.1.0.
> 
> My intention was to test having config schema version that is between two schema versions. In the first iteration of this PR, the migrator would have treated 4.0.1 the same as 4.0.0 and migrated it to 4.1.0. Hope that makes sense.

Now I understand. I've added a new test for missing intermediate steps in the config migration path.

psychedelicious reviewed (changes requested):
Sorry for the delay in returning to this PR. I'm being very cautious of making changes to the config file handling because the it is very sensitive - a small error can turn into a lot of user pain, even if it is easy to fix. So I'm being very picky in this review.

lstein commented on invokeai/app/services/config/config_migrate.py:
```diff
@@ -0,0 +1,89 @@
+        """Define a decorator which registers the migration between two versions."""
+
+        def decorator(function: MigrationFunction) -> MigrationFunction:
+            if any(from_version == m.from_version for m in cls._migrations):
+                raise ValueError(
+                    f"function {function.__name__} is trying to register a migration for version {str(from_version)}, but this migration has already been registered."
+                )
```
Fixed.

lstein commented on tests/test_config.py:
```diff
@@ -265,58 +290,69 @@ def test_get_config_writing(patch_rootdir: None, monkeypatch: pytest.MonkeyPatch
-@pytest.mark.xfail(
-    reason="""
-    This test fails when run as part of the full test suite.
+def test_migration_check() -> None:
```
The unit test `test_migration_check()` checks that the decorator works when presented with contiguous from and to values and fails when there is a discontinuity. However, I have to rework it for the migrator being an instance rather than a class.

lstein replied:
Check it out now.

lstein commented on tests/test_config.py:
```diff
@@ -134,6 +147,16 @@ def test_migrate_v3_backup(tmp_path: Path, patch_rootdir: None):
+    temp_config_file.write_text(v4_config)
+
+    conf = load_and_migrate_config(temp_config_file)
+    assert Version(conf.schema_version) >= Version("4.0.1")
```
This was written so that the test would continue to work when we move to version 4.0.2 -- because the test is running on the "live" config. I'm reworking this as suggested so that the migrator is an instance, in which case the exact value check will work.

lstein commented on invokeai/app/services/config/config_migrate.py:
```diff
@@ -0,0 +1,89 @@
+    function: MigrationFunction
+
+
+class ConfigMigrator:
```
Fixed. `ConfigMigrator` is now an instanced class, and `test_config.py` creates new instances of the migrator for testing decorator functionality, and continuity checking.

The migration steps themselves have been moved into their own module now: `invokeai.app.services.config.migrations`.

I also relocated `get_config()` into `invokeai.app.services.config.config_migrate`, and changed all places that are importing `config_default` to import from `invokeai.app.services.config` instead (`get_config()` is reexported from `__init__.py`) . This avoided circular import problems and makes `config_default.py`  cleaner looking overall.

lstein commented:
@psychedelicious When you have a chance, please give this the once-over. I've converted the migration class into a migration instance, but I'm not sure I did this in the most elegant way.

psychedelicious commented:
Ok, I've revisited this PR and made some changes, simplified a few things and made the tests more rigorous. I'm not entirely convinced the extra complexity in this change is warranted.

I'm running into a problem where the app creates the custom nodes dir & files in the repo root, whenever I run a test. I believe the reason is that my invoke venv is now inside the repo root, and the `find_root` method decides the repo root is the runtime root. Also, I noticed that my live config file is migrated on `main`, when running the config tests.

I added a workaround - setting the root to a temp dir when args weren't parsed. I'm really wary of this change... I'm out of time on this PR for now, I'll revisit later this week.

lstein commented:
> Ok, I've revisited this PR and made some changes, simplified a few things and made the tests more rigorous. I'm not entirely convinced the extra complexity in this change is warranted.
> 
> I'm running into a problem where the app creates the custom nodes dir & files in the repo root, whenever I run a test. I believe the reason is that my invoke venv is now inside the repo root, and the `find_root` method decides the repo root is the runtime root. Also, I noticed that my live config file is migrated on `main`, when running the config tests.
> 
> I added a workaround - setting the root to a temp dir when args weren't parsed. I'm really wary of this change... I'm out of time on this PR for now, I'll revisit later this week.

Blah. This is very bad. We need to set up the test environment with a temporary root directory. For the model manager tests, I created two fixtures like this:

```
@pytest.fixture
def mm2_root_dir(tmp_path_factory) -> Path:
    root_template = Path(__file__).resolve().parent / "data" / "invokeai_root"
    temp_dir: Path = tmp_path_factory.mktemp("data") / "invokeai_root"
    shutil.copytree(root_template, temp_dir)
    return temp_dir

@pytest.fixture
def mm2_app_config(mm2_root_dir: Path) -> InvokeAIAppConfig:
    app_config = InvokeAIAppConfig(models_dir=mm2_root_dir / "models", log_level="info")
    app_config._root = mm2_root_dir
    return app_config
```

Then I use the `mm2_app_config` fixture to get the config rather than calling `get_conf()`. Maybe something like this would work here?

I'll revert to this on Monday to see if I can make the tests behave properly. 

When you say "I'm not sure the extra complexity is worth it," what are you referring to? To me, the current code looks more complex than the original version I wrote using decorators. My goal in the original version was to make it so that you could register a new migration by just adding a single decorated function definition to the `migrations.py` file. In the current code, we have to make code changes in four places: two changes in `migrations.py` to define the migration function and then create a ConfigMigration object, followed two more changes in `config_migrate.py` to import the new ConfigMigration object and then to register it in `load_and_migrate_config`.

I'd like to rethink how this is done. It seems like a lot of work just to remove a key from a dict.

psychedelicious commented:
By "I'm not sure the extra complexity is worth it", I meant having a migration system at all for the config file. A couple simple functions serve the need just fine.

Where I was coming from I added to the PR: We don't have a need to be able to crank out config migrations quickly, but we do have a need for simple and testable code.

It took me a while to understand the intention of ABC, its implementation, the decorator and migration functions. Because the migration functions were only defined inside another function, it was not easy to test each one individually. You had to test the whole migration chain at once. Removing this extra stuff removes a layer of indirection.

psychedelicious commented:
The problem with interactions with the live filesystem aren't caused by tests calling get_config directly. It's that when the app's modules are loaded and classes instantiated in tests, get_config gets called at some point. Another issue is when we read the invocations folder's init, it creates a the custom `nodes` directory. 

My last commit hijacks the root dir when the app isn't launched via script. This does work but I'm not confident in it. This issue is unrelated to the changes in this PR.

psychedelicious commented:
Couple related issues with how tests are running and the config and custom nodes loader:
- `get_config()` is called in several places in the module scope ([example](https://github.com/invoke-ai/InvokeAI/blob/main/invokeai/backend/image_util/depth_anything/__init__.py)). These calls should all be moved inside classes.
- The bulk download images test [imports the whole _live_ app](https://github.com/invoke-ai/InvokeAI/blob/main/tests/app/routers/test_images.py). That test is very thorough but also, a lot of it feels like testing for testing's sake. Fixing the root dir issue should make this OK.
- The custom nodes implementation relies on the `invokeai.app.invocations` module's [init file](https://github.com/invoke-ai/InvokeAI/blob/main/invokeai/app/invocations/__init__.py). In hindsight, this is clearly terrible.

lstein commented:
@psychedelicious At this point I think it's easier just to add ad hoc code to the `load_and_migrate_config()` function. Unless you object, I'm just going to close this PR.

lstein commented:
This looks like a cul-de-sac and I'm closing it.
